### PR TITLE
Add ground atmosphere

### PR DESCRIPTION
**Pull request by bagnell**
* Can be toggled with `Globe.showGroundAtmosphere` and is enabled by default.
* The fade in/out for the night side of the globe can be changes with `Globe.nightFadeOutDistance` and `Globe.nightFadeOutDistance`.
image

### CODE REVIEW

**Comment by hpinkos** [comment]
How does this play with `Globe.enableLighting`?  It looks like we're also defaulting that to `true` now and setting it to `false` seems to have no effect.

**Comment by bagnell** [comment]
If `showGroundAtmosphere` is `true`, then `enableLighting` shows lighting with terrain vertex normals when available. When `showGroundatmosphere` is `false`, it works the same as before.

**Comment by pjcozzi** [comment]
Code looks good at ultra quick glance!

**Comment by bagnell** [comment]
The ground atmosphere is now computed per-fragment in GlobeFS for better lighting of the day/night transition. It is only enabled when the camera is above the `nightFadeOutDistance`, otherwise it is disabled with a `#ifdef`.

This is ready for review.

**Comment by bagnell** [comment]
Bump.

**Comment by lilleyse** [comment]
The default globe looks really good with ground atmosphere but I find the water shading to look a bit washed out. Maybe it needs to be toned down a bit when ground atmosphere is also applied?

Ground atmosphere without shading
globe

Ground atmosphere with shading
globe-with-water

For reference this is without ground atmosphere:

globe-master

**Comment by lilleyse** [comment]
When zooming into the dark side of the globe the lighting transitions are a bit abrupt, going from dark to light to dark:
light-dark

Would this be improved by fading out the ground atmosphere sooner?

**Comment by lilleyse** [comment]
Similar when zooming in on the light side, it gives the impression that the imagery is getting darker when really it is just the ground atmosphere fading away. Fading out the ground atmosphere sooner may help this look better too.

light-dark-2

**Comment by lilleyse** [comment]
The code looks good and 2D and CV follow the same behavior as master.

Do any new tests need to be added?

File: Source/Scene/GlobeSurfaceTileProvider.js [hunk 1]
```diff
@@ -839,6 +840,9 @@ define([
             u_lightingFadeDistance : function() {
                 return this.properties.lightingFadeDistance;
             },
+            u_nightFadeDistance : function() {
```
Comment (pjcozzi) [inline Source/Scene/GlobeSurfaceTileProvider.js]: Minor, but batch this with `u_lightingFadeDistance`?

File: Source/Shaders/GlobeFS.glsl [hunk 2]
```diff
@@ -47,17 +46,21 @@ uniform float u_zoomedOutOceanSpecularIntensity;
 uniform sampler2D u_oceanNormalMap;
 #endif
 
-#ifdef ENABLE_DAYNIGHT_SHADING
+#if defined(ENABLE_DAYNIGHT_SHADING) || defined(GROUND_ATMOSPHERE)
 uniform vec2 u_lightingFadeDistance;
 #endif
 
+#ifdef GROUND_ATMOSPHERE
```
Comment (pjcozzi) [inline Source/Shaders/GlobeFS.glsl]: Ah, maybe not.

**Comment by pjcozzi** [comment]
> Do any new tests need to be added?

Yes, please.

**Comment by pjcozzi** [comment]
Overall looks good at quick glance.

> Similar when zooming in on the light side, it gives the impression that the imagery is getting darker when really it is just the ground atmosphere fading away. Fading out the ground atmosphere sooner may help this look better too.

I have a slight to mild concern that some to many users will not like the washed out look when zoomed out especially with other types of imagery providers such as those from vector sources.  Maybe this is still the best default or maybe there should also be a Sandcastle example showing a few typical configurations, e.g., default realistic use case, vector use case with no atmospheres, realistic use case but fade distance must higher?

**Comment by bagnell** [comment]
> There may be a clean way to separate the ground atmosphere and enable lighting concepts - what if the default setting is `enableLighting=false` and `showGroundAtmosphere=true`, and the ground atmosphere always faces the camera so it looks the same regardless of where on the globe you're looking.

I changed that to be the default.

> Would this be improved by fading out the ground atmosphere sooner?

and 

> Similar when zooming in on the light side, it gives the impression that the imagery is getting darker when really it is just the ground atmosphere fading away. Fading out the ground atmosphere sooner may help this look better too.

I tweaked the fade distances. Give that a try and let me know.

**Comment by bagnell** [comment]
> The default globe looks really good with ground atmosphere but I find the water shading to look a bit washed out. Maybe it needs to be toned down a bit when ground atmosphere is also applied?

I turned down the specular highlight. Did you mean the diffuse as well?

**Comment by bagnell** [comment]
> With the below snippet it seems like vertex normals are not getting shaded as they are in master.

I think this was a mistake when pulled from the HDR branch. Fixed now, but, note to self, this is also probably a problem in that branch.

**Comment by pjcozzi** [comment]
What's left for now?  Let's please get this over the finish line before pushing on IBL.

**Comment by lilleyse** [comment]
> I turned down the specular highlight. Did you mean the diffuse as well?

Actually the specular highlight might have been okay, mainly it is the diffuse that seems too strong:

bright

**Comment by lilleyse** [comment]
This is a good example of what @pjcozzi means with looking washed out on different types of imagery.

vector-lighting-off
vector-lighting-on

**Comment by lilleyse** [comment]
I also agree that a "Globe Lighting" Sandcastle that has toggles for
* Ground Atmosphere
* Enable lighting
* Switch between shading with terrain vertex normals or not
* Water shading
* Sliders for controlling the fade in and fade out distances

would be pretty useful to have generally, especially as this change rolls out and people are interested in customizing it. We could just point them to the demo.

**Comment by lilleyse** [comment]
I tweaked some of the parameters for the fade in - it's a little smoother now and looks a bit better when `enableLighting` is on. https://github.com/AnalyticalGraphicsInc/cesium/pull/6877/commits/e1005febe50eeb4bde2befeaea5b054badc9baa2

Though setting `nightFadeOutDistance` didn't seem to make a difference even when set to 0. Is there some hidden minimum?

fade-in

**Comment by bagnell** [comment]
> Though setting `nightFadeOutDistance` didn't seem to make a difference even when set to 0. Is there some hidden minimum?

The `nightFadeOutDistance` is the distance until it fades to a fully lit atmosphere. So if you set it less than `lightingFadeOutDistance` you won't see a difference. Set it to something greater than `lightingFadeOutDistance` but less than `nightFadeInDistance` and you'll see what I mean.

**Comment by bagnell** [comment]
@lilleyse This is ready for another look.

**Comment by lilleyse** [comment]
Looks really good - the water shading from far out is now much nicer now! I added a few last tweaks. Mainly just upped the specular brightness and changed the night fade in.

There is one bug in the atmosphere demo where clicking the reset button doesn't seem to update the sliders.

Does anyone else want to check this out before we merge? Demo of this branch here:

http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/atmosphere/Apps/Sandcastle/index.html?src=Atmosphere.html

**Comment by pjcozzi** [comment]
Looks great at quick glance!

**Comment by bagnell** [comment]
> There is one bug in the atmosphere demo where clicking the reset button doesn't seem to update the sliders.

Fixed.

File: Apps/Sandcastle/gallery/Atmosphere.html [hunk 3]
```diff
@@ -154,6 +154,14 @@
...
     globe.nightFadeOutDistance = defaultNightFadeOut;
     globe.nightFadeInDistance = defaultNightFadeIn;
+
+    document.getElementById('lightingFadeOutDistanceSlider').value = defaultLightFadeOut;
+    document.getElementById('lightingFadeOutDistanceText').value = defaultLightFadeOut;
+    document.getElementById('lightingFadeInDistanceSlider').value = defaultLightFadeIn;
+    document.getElementById('lightingFadeInDistanceText').value = defaultLightFadeIn;
+    document.getElementById('nightFadeOutDistanceSlider').value = defaultNightFadeOut;
+    document.getElementById('nightFadeOutDistanceText').value = defaultNightFadeOut;
+    document.getElementById('nightFadeInDistanceSlider').value = defaultNightFadeOut;
```
Comment (lilleyse) [inline Apps/Sandcastle/gallery/Atmosphere.html]: I think the proper way to do this in knockout is:

```
viewModel.lightingFadeOutDistance = defaultLightFadeOut;
viewModel.lightingFadeInDistance = defaultLightFadeIn;
viewModel.nightFadeOutDistance = defaultNightFadeOut;
viewModel.nightFadeInDistance = defaultNightFadeIn;
```

Comment (bagnell) [reply Apps/Sandcastle/gallery/Atmosphere.html]: Thanks. Updated.

**Comment by bagnell** [comment]
I also noticed that the fog was broken, but it's now fixed. This is ready for another look.

**Comment by lilleyse** [comment]
Looks good to me. Thanks @bagnell, this is a really nice visual quality improvement!

**Comment by lilleyse** [comment]
Just added some last bits of documentation: https://github.com/AnalyticalGraphicsInc/cesium/pull/6877/commits/b75d1cfb6a4479069a3ec4bd450a3bb547f8138a

**Pull request merged**